src/Hosting/Hosting/src/Internal/HostingApplication.cs

```diff
@@ -122,13 +122,26 @@ internal class Context
         {
             public HttpContext? HttpContext { get; set; }
             public IDisposable? Scope { get; set; }
-            public Activity? Activity { get; set; }
+            public Activity? Activity
+            {
+                get => HttpActivityFeature?.Activity;
+                set
+                {
+                    if (HttpActivityFeature is null)
+                    {
+                        HttpActivityFeature = new ActivityFeature(value!);
+                    }
+                    HttpActivityFeature.Activity = value!;
```

Comment:
What happens if HostingApplication.Activity is set with a null activity?

Reply:
It isn't https://github.com/dotnet/aspnetcore/pull/31180/files#diff-f4b6e9bc19da99141d4427d88673ed12fdf3474082f571b027a583cb6959ca99R62

Reply:
I guess we can add an assert here

Reply:
+1 assert.

Reply:
Oh, I guess we do set it to null in `Reset()`. But we don't want to null out the feature since we are trying to avoid allocations.